File: TsinghuaNet.Helper/ViewModels/NetViewModel.vb
Imports System.ComponentModel
Imports System.Windows.Input

Public MustInherit Class NetViewModel
    Inherits NetObservableBase

    Public Sub New()
        LoadSettings()
        AddHandler Credential.PropertyChanged, AddressOf OnCredentialStateChanged
    End Sub

    Public MustOverride Sub LoadSettings()

    Public MustOverride Sub SaveSettings()

    Private Sub OnCredentialStateChanged(sender As Object, e As PropertyChangedEventArgs)
        If e.PropertyName = "State" Then
            Refresh()
        End If
    End Sub

    Private _Status As INetStatus
    Public Property Status As INetStatus
        Get
            Return _Status
        End Get
        Set(value As INetStatus)
            SetProperty(_Status, value, onChanged:=AddressOf RefreshStatus)
        End Set
    End Property

    Private _SuggestState As NetState
    Public Property SuggestState As NetState
        Get
            Return _SuggestState
        End Get
        Set(value As NetState)
            SetProperty(_SuggestState, value, onChanged:=AddressOf OnSuggestStateChanged)
        End Set
    End Property
    Protected Overridable Sub OnSuggestStateChanged()

    End Sub

    Public ReadOnly Property RefreshStatusCommand As ICommand = New Command(Me, AddressOf RefreshStatus)
    Public Async Sub RefreshStatus()
        Await RefreshStatusAsync()
    End Sub
    Public Async Function RefreshStatusAsync() As Task
        Try
            IsBusy = True
            Await Status.RefreshAsync()
            SuggestState = Await Status.SuggestAsync()
        Finally
            IsBusy = False
        End Try
    End Function

    Private _OnlineUser As FluxUser
    Public Property OnlineUser As FluxUser
        Get
            Return _OnlineUser
        End Get
        Set(value As FluxUser)
            SetProperty(_OnlineUser, value)
        End Set
    End Property

    Public Event ReceivedResponse As EventHandler(Of LogResponse)

    Protected Overridable Sub OnReceivedResponse(res As LogResponse)
        RaiseEvent ReceivedResponse(Me, res)
    End Sub

    Friend Async Function NetCommandExecuteAsync(executor As Func(Of IConnect, Task(Of LogResponse))) As Task
        Try
            IsBusy = True
            Dim helper = Credential.GetHelper()
            OnReceivedResponse(Await executor(helper))
        Catch ex As Exception
            OnReceivedResponse(New LogResponse(False, ex.Message))
        Finally
            IsBusy = False
        End Try
    End Function

    Public ReadOnly Property LoginCommand As ICommand = New NetCommand(Me, AddressOf LoginAsync)
    Protected Overridable Async Function LoginAsync(helper As IConnect) As Task(Of LogResponse)
        Dim res As New LogResponse(True, "登录成功")
        If helper IsNot Nothing Then
            Dim r = Await helper.LoginAsync()
            If Not r.Succeed Then
                res = r
            End If
        End If
        Await RefreshAsync(helper)
        Return res
    End Function
    Public Function LoginAsync() As Task
        Return NetCommandExecuteAsync(AddressOf LoginAsync)
    End Function
    Public Async Sub Login()
        Await LoginAsync()
    End Sub

    Public ReadOnly Property LogoutCommand As ICommand = New NetCommand(Me, AddressOf LogoutAsync)
    Protected Overridable Async Function LogoutAsync(helper As IConnect) As Task(Of LogResponse)
        Dim res As New LogResponse(True, "注销成功")
        If helper IsNot Nothing Then
            Dim r = Await helper.LogoutAsync()
            If Not r.Succeed Then
                res = r
            End If
        End If
        Await RefreshAsync(helper)
        Return res
    End Function
    Public Function LogoutAsync() As Task
        Return NetCommandExecuteAsync(AddressOf LogoutAsync)
    End Function
    Public Async Sub Logout()
        Await LogoutAsync()
    End Sub

    Public ReadOnly Property RefreshCommand As ICommand = New NetCommand(Me, AddressOf RefreshAsync)
    Protected Overridable Async Function RefreshAsync(helper As IConnect) As Task(Of LogResponse)
        Dim user As FluxUser = Nothing
        If helper IsNot Nothing Then
            user = Await helper.GetFluxAsync()
        End If
        OnlineUser = user
        Return New LogResponse(True, String.Empty)
    End Function
    Public Function RefreshAsync() As Task
        Return NetCommandExecuteAsync(AddressOf RefreshAsync)
    End Function
    Public Async Sub Refresh()
        Await RefreshAsync()
    End Sub

    Private _AutoLogin As Boolean
    Public Property AutoLogin As Boolean
        Get
            Return _AutoLogin
        End Get
        Set(value As Boolean)
            SetProperty(_AutoLogin, value)
        End Set
    End Property

    Private _EnableFluxLimit As Boolean
    Public Property EnableFluxLimit As Boolean
        Get
            Return _EnableFluxLimit
        End Get
        Set(value As Boolean)
            SetProperty(_EnableFluxLimit, value)
        End Set
    End Property

    Private _FluxLimit As ByteSize
    Public Property FluxLimit As ByteSize
        Get
            Return _FluxLimit
        End Get
        Set(value As ByteSize)
            SetProperty(_FluxLimit, value)
        End Set
    End Property
End Class

File: TsinghuaNet/Models/NetViewModel.cs
using System;
using System.ComponentModel;
using System.Threading.Tasks;
using System.Windows.Input;

namespace TsinghuaNet.Models
{
    public abstract class MainViewModelBase : NetViewModelBase
    {
        public MainViewModelBase()
        {
            LoadSettings();
            RefreshStatusCommand = new Command(this, RefreshStatus);
            LoginCommand = new NetCommand(this, LoginAsync);
            LogoutCommand = new NetCommand(this, LogoutAsync);
            RefreshCommand = new NetCommand(this, RefreshAsync);
            Credential.PropertyChanged += OnCredentialStateChanged;
        }

        public abstract void LoadSettings();

        public abstract void SaveSettings();

        private void OnCredentialStateChanged(object sender, PropertyChangedEventArgs e)
        {
            if (e.PropertyName == "State")
                Refresh();
        }

        public INetStatus Status { get; set; }
        private void OnStatusChanged() => RefreshStatus();

        public ICommand RefreshStatusCommand { get; }
        public async void RefreshStatus() => await RefreshStatusAsync();
        public async Task RefreshStatusAsync()
        {
            try
            {
                IsBusy = true;
                await Status.RefreshAsync();
                Credential.State = await Status.SuggestAsync();
            }
            finally
            {
                IsBusy = false;
            }
        }

        public FluxUser OnlineUser { get; set; }

        public TimeSpan OnlineTime { get; set; }

        public string Response { get; set; }

        public event EventHandler<LogResponse> ReceivedResponse;

        protected virtual void OnReceivedResponse(LogResponse res) => ReceivedResponse?.Invoke(this, res);

        internal async Task NetCommandExecuteAsync(Func<IConnect, Task<LogResponse>> executor)
        {
            try
            {
                IsBusy = true;
                var helper = Credential.GetHelper();
                OnReceivedResponse(await executor(helper));
            }
            catch (Exception ex)
            {
                OnReceivedResponse(new LogResponse(false, ex.Message));
            }
            finally
            {
                IsBusy = false;
            }
        }

        public ICommand LoginCommand { get; }
        protected virtual async Task<LogResponse> LoginAsync(IConnect helper)
        {
            LogResponse res = new LogResponse(true, "登录成功");
            if (helper != null)
            {
                var r = await helper.LoginAsync();
                if (!r.Succeed)
                    res = r;
            }
            await RefreshAsync(helper);
            return res;
        }
        public Task LoginAsync() => NetCommandExecuteAsync(LoginAsync);
        public async void Login() => await LoginAsync();

        public ICommand LogoutCommand { get; }
        protected virtual async Task<LogResponse> LogoutAsync(IConnect helper)
        {
            LogResponse res = new LogResponse(true, "注销成功");
            if (helper != null)
            {
                var r = await helper.LogoutAsync();
                if (!r.Succeed)
                    res = r;
            }
            await RefreshAsync(helper);
            return res;
        }
        public Task LogoutAsync() => NetCommandExecuteAsync(LogoutAsync);
        public async void Logout() => await LogoutAsync();

        public ICommand RefreshCommand { get; }
        protected virtual async Task<LogResponse> RefreshAsync(IConnect helper)
        {
            FluxUser user = default;
            if (helper != null)
                user = await helper.GetFluxAsync();
            OnlineUser = user;
            return new LogResponse(true, string.Empty);
        }
        public Task RefreshAsync() => NetCommandExecuteAsync(RefreshAsync);
        public async void Refresh() => await RefreshAsync();
    }
}
